aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,575 @@
+        ciphertext: &mut [u8],
+        error_value: E,
+        kdf: F,
+    ) -> Result<R, E>
+    where
+        F: FnOnce(&[u8]) -> Result<R, E>,
+    {
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) });
+        let ctx = if let Ok(ctx) = ctx {
+            ctx
+        } else {
+            return Err(error_value);
+        };
+
+        if 1 != unsafe {
+            EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            )
+        } {
+            return Err(error_value);
+        }
+
+        kdf(&shared_secret)
+    }
+
+    /// Creates a new KEM private key from raw bytes. This method is NOT meant to generate
+    /// a new private key, rather it restores a `KemPrivateKey` that was previously converted
+    /// to raw bytes.
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    pub fn new(alg: &'static Algorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.secret_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        let privkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_secret_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(PrivateKey {
+            algorithm: alg,
+            evp_pkey: privkey,
+            priv_key: bytes.into(),
+        })
+    }
+}
+
+impl Drop for PrivateKey {
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl AsRef<[u8]> for PrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+#[derive(Debug)]
+pub struct PublicKey {
+    algorithm: &'static Algorithm,
+    evp_pkey: LcPtr<EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl PublicKey {
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the encapsulation is finished, `encapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
```

Comment:
do you have a reference for this definition?

Reply:
Rust is weird, I assume you are looking for the function definition of `kdf` argument? It's a generic parameter define in the function definition's where clause.

```
where
    F: FnOnce(&[u8], &[u8]) -> Result<R, E>
```